x/gamm/pool-models/balancer/amm.go

```diff
@@ -281,39 +283,101 @@ func (p *Pool) CalcJoinPoolShares(_ sdk.Context, tokensIn sdk.Coins, swapFee sdk
-		poolAsset.Token.Amount = poolAssetsByDenom[coin.Denom].Token.Amount.Add(coin.Amount)
-		poolAssetsByDenom[coin.Denom] = poolAsset
-	}
-
-	newTotalShares := totalShares.Add(numShares)
+	// This is new liquidity that to be added to the pool from exact ratio join.
+	// If there is remainingLiqudity, we will add it to newLiquidity by
+	// performing single asset join on each coin in remainingLiquidity.
+	newLiquidity = tokensIn.Sub(remainingLiquidity)
 
 	// If there are coins that couldn't be perfectly joined, do single asset joins
 	// for each of them.
-	if !remCoins.Empty() {
-		for _, coin := range remCoins {
-			newShares, err := p.calcSingleAssetJoin(coin, swapFee, poolAssetsByDenom[coin.Denom], newTotalShares)
-			if err != nil {
-				return sdk.ZeroInt(), sdk.NewCoins(), err
-			}
-
-			newLiquidity = newLiquidity.Add(coin)
-			newTotalShares = newTotalShares.Add(newShares)
-			numShares = numShares.Add(newShares)
+	if !remainingLiquidity.Empty() {
+		// update pool assets for accurate calcSingleAssetJoin calculation
+		if err := updateIntermediaryPoolAssets(newLiquidity, poolAssetsByDenom); err != nil {
+			return sdk.ZeroInt(), sdk.NewCoins(), err
 		}
+
+		// update total shares for accurate calcSingleAssetJoin calculation
+		newTotalShares := totalShares.Add(numShares)
+
+		newNumSharesFromRemaining, newLiquidityFromRemaining, err := p.calcJoinSingleAssetTokensIn(remainingLiquidity, newTotalShares, poolAssetsByDenom, swapFee)
+		if err != nil {
+			return sdk.ZeroInt(), sdk.NewCoins(), err
+		}
+		numShares.Add(newNumSharesFromRemaining)
+		newLiquidity = newLiquidity.Add(newLiquidityFromRemaining...)
 	}
 
 	return numShares, newLiquidity, nil
 }
 
+// getPoolAssetsByDenom return a mapping from pool asset
+// denom to the pool asset itself. There must be no duplicates.
+// Returns error, if any found.
+func getPoolAssetsByDenom(poolAssets []PoolAsset) (map[string]PoolAsset, error) {
+	poolAssetsByDenom := make(map[string]PoolAsset)
+	for _, poolAsset := range poolAssets {
+		_, ok := poolAssetsByDenom[poolAsset.Token.Denom]
+		if ok {
+			return nil, fmt.Errorf(errMsgFormatRepeatingPoolAssetsNotAllowed, poolAsset.Token.Denom)
+		}
+
+		poolAssetsByDenom[poolAsset.Token.Denom] = poolAsset
+	}
+	return poolAssetsByDenom, nil
+}
+
+// updateIntermediaryPoolAssets updates poolAssetsByDenom with liquidity.
+//
+// all liqidity coins must exist in poolAssetsByDenom. Returns error, if not.
+//
+// This is a helper function that is useful for updating the pool asset amounts
+// as an intermediary step in a multi-join methods such as CalcJoinPoolShares.
+// In CalcJoinPoolShares with multi-asset joins, we first attempt to do
+// a MaximalExactRatioJoin that might leave out some tokens in.
+// Then, for every remaining tokens in, we attempt to do a single asset join.
+// Since the first step (MaximalExactRatioJoin) affects the pool liqudity due to slippage,
+// we would like to account for that in the subsequent steps of single asset join
```

Comment:
Do you mean that the caller will have to update the internal state of the pool, meaning that this function is non-mutative?

Wouldn't you think that is implied since this is a function (not a method) and does not have access to the pool state?

Reply:
Nvm, I see what you mean

Reply:
The comment already mentions that on the first line: https://github.com/osmosis-labs/osmosis/blob/9c1a2da3e0a9b875680022f4a77c5daf65d6d294/x/gamm/pool-models/balancer/amm.go#L334